src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -135,40 +152,88 @@ public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams reques
+            Diagnostic[] unmappedDiagnostics,
+            RazorCodeDocument codeDocument,
+            DocumentSnapshot documentSnapshot)
+        {
+            var sourceText = await documentSnapshot.GetTextAsync();
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var classifiedSpans = syntaxTree.GetClassifiedSpans();
+
+            var processedAttributes = new Dictionary<TextSpan, bool>();
+
+            var filteredDiagnostics = unmappedDiagnostics
+                .Where(d =>
+                    // TODO: undesired HTML Diagnostic codes
+                    // (https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401) ||
+                    !InAttributeContainingCSharp(d, sourceText, syntaxTree, classifiedSpans, processedAttributes))
+                .ToArray();
+
+            return filteredDiagnostics;
+
+            static bool InAttributeContainingCSharp(
+                Diagnostic d,
+                SourceText sourceText,
+                RazorSyntaxTree syntaxTree,
+                IReadOnlyList<ClassifiedSpanInternal> classifiedSpans,
+                Dictionary<TextSpan, bool> processedAttributes)
             {
-                return kind == RazorLanguageKind.CSharp ?
-                    CanCSharpDiagnosticBeFiltered(d) :
-                    CanHTMLDiagnosticBeFiltered(d);
+                // Examine the _end_ of the diagnostic to see if we're at the
+                // start of an (im/ex)plicit expression. Looking at the start
+                // of the diagnostic isn't sufficient.
+                var absoluteIndex = d.Range.End.GetAbsoluteIndex(sourceText);
+                var change = new SourceChange(absoluteIndex, 0, string.Empty);
+                var owner = syntaxTree.Root.LocateOwner(change);
+
+                var markupAttributeNode = owner.FirstAncestorOrSelf<MarkupAttributeBlockSyntax>();
+
+                if (markupAttributeNode != null)
+                {
+                    if (!processedAttributes.TryGetValue(markupAttributeNode.FullSpan, out var doesAttributeContainsCSharp))
+                    {
+                        doesAttributeContainsCSharp = CheckIfAttributeContainsCSharp(markupAttributeNode);
+                        processedAttributes.Add(markupAttributeNode.FullSpan, doesAttributeContainsCSharp);
+                    }
+
+                    return doesAttributeContainsCSharp;
+                }
+
+                return false;
             }
 
-            static bool CanCSharpDiagnosticBeFiltered(Diagnostic d) =>
-                (DiagnosticsToIgnore.Contains(d.Code?.String) &&
-                 d.Severity != DiagnosticSeverity.Error);
+            static bool CheckIfAttributeContainsCSharp(MarkupAttributeBlockSyntax attributeNode)
+            {
+                var containsNonHTMLNodes = attributeNode.DescendantNodes()
+                    .Any(n =>
+                        n is CSharpCodeBlockSyntax ||
```

Comment:
Hmm, why check for C# of checking for !Html?

Reply:
From what I could tell there wasn't a clear way for us to do a !HTML (ex. `n is MarkupBlockSyntax || n is MarkupSyntaxNode` wouldn't be sufficient) as there are many different types of nodes which wouldn't fall in those categories in the descendants list, but we wouldn't want to exclude. Ex. basic syntax/punctuation ([`SyntaxNode.SyntaxKind.Colon`](https://github.com/dotnet/aspnetcore/blob/master/src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/SyntaxKind.cs#L64)). Special casing all of those various types would bloat this quite a bit, so I ended up taking the !C#/!Razor comment approach.

Please do let me know if there's a concise way we could do !HTML.

Reply:
The the SyntaxNode.SyntaxKind.COlon is representative of a "Token", (`node.IsToken`). Tokens can be utilized in C# or HTML content. So if you ignore bottom level tokens and only look at their parents I believe there's a well formed parent node you can interoperate on. For reference here's the applicable nodes in the Razor syntax: https://github.com/dotnet/aspnetcore/blob/master/src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/Syntax.xml

Reply:
Thanks that did it!

![image](https://user-images.githubusercontent.com/14852843/105392413-2ae39e80-5be9-11eb-9940-b68b77e871fb.png)